mediatek: Add support for Wavlink WL-WN586X3

themaverickdm opened the pull request:
Hardware
--------
- SOC: MediaTek MT7981
- ram: 256MB DDR3
- FLASH: 16MB SPI-NOR
- Ethernet: 2x1Gb Lan 1x1Gb Wan
- WIFI: MediaTek MT7981 2x2 DBDC 802.11ax 2T2R (2.4/5)
- LEDs: 2xLan 1x Wan  1x WIFI 1xSTATUS

MAC table, same as stock firmware:
LAN: 80:3F:5D:xx:xx:x1 partition "hw" at 0x44e (ASCII)
WAN: 80:3F:5D:xx:xx:x2 partition "hw" at 0x460 (ASCII)
 2G: 80:3F:5D:xx:xx:x3 partition "factory" at 0x4 (binary), on label
 5G: 80:3F:5D:xx:xx:x3 Same as 2G

Installation Method 1: ssh
--------------------------
1. Connect PC to the lan port. Set the PC IP to 192.168.10.100 if required.
2. Navigate to http://192.168.10.1/
3. Log into the Wavlink WebGUI. Default username/password is admin/admin.
4. Use WebGUI to upgrade the firmware to WAVLINK_WN586X3-A_M86X3A_V240113_WO-GDBYFM-modified.bin downloaded from https://github.com/themaverickdm/firmware-misc/tree/main/wavlink/wl-wn586x3 Warning: All settings will be lost!
5. Wait about 5 minutes, and after flashing is completed, log into the router using (with admin123 as password):
ssh root@192.168.10.1
6. scp the openwrt image file onto the router, usually under /tmp somewhere.
openwrt-mediatek-filogic-wavlink_wl-wn586x3-squashfs-sysupgrade.bin
7. Flash openwrt image file like so:
mtd write openwrt-mediatek-filogic-wavlink_wl-wn586x3-squashfs-sysupgrade.bin firmware
Warning: Previous firmware will be overwritten!
8. Wait about 5 minutes, and after the flashing is completed, set the PC IP to 192.168.1.100 if required and log into the router like so:
ssh root@192.168.1.1

Installation Method 2: u-boot
-----------------------------
1. Connect UART:  TX-> 586X3 RX, RX-> 586X3 TX, GND-> 586 GND.
2. Connect PC to the wan (not lan!) port.
3. Setup the tftp server on PC, set IP to 192.168.10.100,
4. Power on the device. Select '2' to upgrade firmware in Uboot.
5. Input the image name and start to upgrade.

Uboot console log:

CPU:   MediaTek MT7981
Model: mt7981-rfb
DRAM:  256 MiB
Core:  34 devices, 13 uclasses, devicetree: embed
Loading Environment from nowhere... OK
In:    serial@11002000
Out:   serial@11002000
Err:   serial@11002000
Net:
Warning: ethernet@15100000 (eth0) using random MAC address -
02:47:fb:b2:53:2d
eth0: ethernet@15100000
UBOOT WN586X3A
gpio: pin 9 (gpio 9) value is 0
gpio: pin 10 (gpio 10) value is 0
gpio: pin 5 (gpio 5) value is 0
gpio: pin 12 (gpio 12) value is 0
gpio: pin 13 (gpio 13) value is 0

  *** U-Boot Boot Menu ***

      1. Startup system (Default)
      2. Upgrade firmware
      3. Upgrade ATF BL2
      4. Upgrade ATF FIP
      5. Upgrade single image
      6. Load image
      0. U-Boot console

themaverickdm commented:
Hello,
This change is derived from https://github.com/openwrt/openwrt/pull/13783 with changes as suggested by the reviewers and label mac address.
Note that this commit by itself is not sufficient for full support of WL-WN586X3. Kernel patches adding support for NOR SPI device BoHong bh25q128as are required as well. These will come soon.
Not affiliated with Wavlink - just want my device to work with openwrt.
Thanks.

themaverickdm commented:
@csharper2005, @raenye, @musashino205, would you all please review this PR?
Thanks!

raenye commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,262 @@
+};
+
+&mdio_bus {
+	switch: switch@0 {
```
```suggestion
	switch: switch@1f {
```

raenye commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,262 @@
+			gpios = <&pio 10 GPIO_ACTIVE_LOW>;
+		};
+
+		led-3{
```
```suggestion
		led-3 {
```

raenye commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,262 @@
+			gpios = <&pio 5 GPIO_ACTIVE_LOW>;
+		};
+
+		led-1{
```
```suggestion
		led-1 {
```

raenye commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,262 @@
+		compatible = "gpio-leds";
+
+		led-0 {
+			label = "blue:wan";
```
Deprecated syntax for LEDs; please use `color` and `function`.

themaverickdm replied:
Thanks for your review.
I have made the changes that you have suggested, including the syntax for LEDs as far as I understand how they are supposed to be. Please check, and let me know how it looks.
Thanks.

raenye commented:
I don't see any change related to the `label` property of LEDs...

themaverickdm commented:
> I don't see any change related to the `label` property of LEDs...

Sorry for being thick, but would you give me some examples of what is expected? Thanks.

themaverickdm commented:
> I don't see any change related to the `label` property of LEDs...

Please check the latest changes. Thanks!

raenye commented:
Good, now just drop the `label` property altogether, add `function-enumerator` to differentiate LAN 1 from 2, and fix the references to the (slightly modified) LED names in `01_leds`.
See, e.g., https://github.com/openwrt/openwrt/commit/485adc9d3c436d31f9713a7d8d84adf266754e26

themaverickdm commented:
> Good, now just drop the `label` property altogether, add `function-enumerator` to differentiate LAN 1 from 2, and fix the references to the (slightly modified) LED names in `01_leds`. See, e.g., [485adc9](https://github.com/openwrt/openwrt/commit/485adc9d3c436d31f9713a7d8d84adf266754e26)

Done. Please check, and let me know. Thanks.

raenye commented:
LGTM.
Bye, what makes the WLAN LED 5G (hardware wise)? Is there any marking, or you could just drop the function enumerator from it?

Also, consider setting this LED via linux trigger directly from the DTS, instead of going through 01_leds.

musashino205 commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,270 @@
+					};
+				};
+			};
+
```
please remove this empty line

musashino205 commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,270 @@
+				label = "firmware";
+				reg = <0x180000 0xe00000>;
+			};
+			partition@f80000 {
```
please insert a new line before ```partition@f80000 {```

musashino205 commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,270 @@
+		phy-mode = "2500base-x";
+		nvmem-cells = <&macaddr_hw_44e 0>;
+		nvmem-cell-names = "mac-address";
+		fixed-link {
```
please insert a new line before ```fixed-link {```

csharper2005 commented on target/linux/mediatek/dts/mt7981b-wavlink-wl-wn586x3.dts:
```diff
@@ -0,0 +1,270 @@
+
+		port@6 {
+			reg = <6>;
+			label = "cpu";
```
cpu label can be dropped.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -47,6 +47,12 @@ routerich,ax3000)
+	ucidef_set_led_netdev "lan-1" "lan-1" "blue:lan-1" "lan1" "link tx rx"
+	ucidef_set_led_netdev "lan-2" "lan-2" "blue:lan-2" "lan2" "link tx rx"
+	ucidef_set_led_netdev "wan" "wan" "blue:wan" "eth1" "link tx rx"
+	ucidef_set_led_netdev "wlan5g" "WLAN5G" "blue:wlan-50" "phy1-ap0" "link tx rx"
```
I suggest to add `linux,default-trigger = "phy0tpt";` to the led-1 description instead of this line.

themaverickdm replied:
Done in latest change.

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -869,6 +869,17 @@ define Device/ubnt_unifi-6-plus
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_DTS_LOADADDR := 0x47000000
+  IMAGE_SIZE := 15424k
+  DEVICE_PACKAGES := kmod-mt7981-firmware
```
I wanted to suggest adding `mt7981-wo-firmware` too. But this is spi and I'm not sure...

themaverickdm replied:
Please let me know if I should add this. Thanks.

csharper2005 replied:
wo-firmware can be installed by the end user, so let's keep this as is.

P.S. You can add `-uboot-envtools` to save a bit more free space.

hauke replied:
I think you should add `mt7981-wo-firmware`. It is needed to offload wifi and we add it for the other mt7981 devices.

themaverickdm replied:
Done.

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -869,6 +869,17 @@ define Device/ubnt_unifi-6-plus
 TARGET_DEVICES += ubnt_unifi-6-plus
 
+define Device/wavlink_wl-wn586x3
+  DEVICE_VENDOR := Wavlink
```
I suggest use caps there - WAVLINK (the same as in dts file).

themaverickdm commented:
> LGTM. BTW, what makes the WLAN LED 5G (hardware wise)? Is there any marking, or you could just drop the function enumerator from it?
> 
> Also, consider setting this LED via linux trigger directly from the DTS, instead of going through 01_leds.

The LED is marked as "WiFi". The original author tied it to 5G, so I just used that. Added trigger for this LED in the latest chnage.

themaverickdm commented:
Hello, if everything looks good, when can this change be merged? Thanks.

themaverickdm commented:
Hello, what else do I need to do to get this change merged? Do I need to ask for additional reviewers? Thanks.

rmilecki commented:
@themaverickdm it seems that this is heavily based on https://github.com/openwrt/openwrt/pull/13783 but you claim to be the author. I don't think it's legally acceptable.

In my opinion you should:
1. Keep @huangsijiahsj as a git commit author and keep his `Signed-off-by`
2. Add tags `Co-authored-by` and `Signed-off-by` tags with your name and ideally make it full one (extend "R" to a full name)

themaverickdm commented:
> @themaverickdm it seems that this is heavily based on #13783 but you claim to be the author. I don't think it's legally acceptable.
> 
> In my opinion you should:
> 
> 1. Keep @huangsijiahsj as a git commit author and keep his `Signed-off-by`
> 2. Add tags `Co-authored-by` and `Signed-off-by` tags with your name and ideally make it full one (extend "R" to a full name)

Interesting, as I did not know this. I did mention that it was based on #13783, but there have been quite a few changes since then.

Anyway, not an issue. I will upload the changes that you suggested soon.

themaverickdm commented:
@rmilecki, I was not able to apply #13783 at all, so I had no choice but to just add the **Co-authored-by** and **Signed-off-by** tags with @huangsijiahsj.
Come to think of it, I was not able to apply #13783 even when I first started working on supporting WL-WN586X3. Also, @huangsijiahsj had closed #13783 without getting it merged, which I took to mean that they were not interested in working on it anymore.

Commit:
mediatek: filogic: Add support for Wavlink WL-WN586X3

Hardware
--------
- SOC: MediaTek MT7981
- ram: 256MB DDR3
- FLASH: 16MB SPI-NOR
- Ethernet: 2x1Gb Lan 1x1Gb Wan
- WIFI: MediaTek MT7981 2x2 DBDC 802.11ax 2T2R (2.4/5)
- LEDs: 2xLan 1x Wan 1x WIFI 1xSTATUS

MAC table, same as stock firmware:
LAN: 80:3F:5D:xx:xx:x1 partition "hw" at 0x44e (ASCII)
WAN: 80:3F:5D:xx:xx:x2 partition "hw" at 0x460 (ASCII)
 2G: 80:3F:5D:xx:xx:x3 partition "factory" at 0x4 (binary), on label
 5G: 80:3F:5D:xx:xx:x3 Same as 2G

Installation Method 1: ssh
--------------------------
1. Connect PC to the lan port. Set the PC IP to 192.168.10.100 if
   required.
2. Navigate to http://192.168.10.1/
3. Log into the Wavlink WebGUI. Default username/password is
   admin/admin.
4. Use WebGUI to upgrade the firmware to
   WAVLINK_WN586X3-A_M86X3A_V240113_WO-GDBYFM-modified.bin
   downloaded from
   https://github.com/themaverickdm/firmware-misc/tree/main/wavlink/wl-wn586x3
   Warning: All settings will be lost!
5. Wait about 5 minutes, and after flashing is completed, log into
   the router using (with admin123 as password):
   ssh root@192.168.10.1
6. scp the openwrt image file onto the router, usually under /tmp
   somewhere.
   openwrt-mediatek-filogic-wavlink_wl-wn586x3-squashfs-sysupgrade.bin
7. Flash openwrt image file like so:
   mtd write \
     openwrt-mediatek-filogic-wavlink_wl-wn586x3-squashfs-sysupgrade.bin \
     firmware
   Warning: Previous firmware will be overwritten!
8. Wait about 5 minutes, and after the flashing is completed, set
   the PC IP to 192.168.1.100 if required and log into the router
   like so:
   ssh root@192.168.1.1

Installation Method 2: u-boot
-----------------------------
1. Connect UART:  TX-> 586X3 RX, RX-> 586X3 TX, GND-> 586 GND.
2. Connect PC to the wan (not lan!) port.
3. Setup the tftp server on PC, set IP to 192.168.10.100,
4. Power on the device. Select '2' to upgrade firmware in Uboot.
5. Input the image name and start to upgrade.

Uboot console log:

CPU:   MediaTek MT7981
Model: mt7981-rfb
DRAM:  256 MiB
Core:  34 devices, 13 uclasses, devicetree: embed
Loading Environment from nowhere... OK
In:    serial@11002000
Out:   serial@11002000
Err:   serial@11002000
Net:
Warning: ethernet@15100000 (eth0) using random MAC address -
02:47:fb:b2:53:2d
eth0: ethernet@15100000
UBOOT WN586X3A
gpio: pin 9 (gpio 9) value is 0
gpio: pin 10 (gpio 10) value is 0
gpio: pin 5 (gpio 5) value is 0
gpio: pin 12 (gpio 12) value is 0
gpio: pin 13 (gpio 13) value is 0

  *** U-Boot Boot Menu ***

      1. Startup system (Default)
      2. Upgrade firmware
      3. Upgrade ATF BL2
      4. Upgrade ATF FIP
      5. Upgrade single image
      6. Load image
      0. U-Boot console

Co-authored-by: R Maru <deviantmaru@gmail.com>
Signed-off-by: R Maru <deviantmaru@gmail.com>
Signed-off-by: Sijia Huang <engineer31@win-star.com>

themaverickdm commented:
> @themaverickdm it seems that this is heavily based on #13783 but you claim to be the author. I don't think it's legally acceptable.
> 
> In my opinion you should:
> 
> 1. Keep @huangsijiahsj as a git commit author and keep his `Signed-off-by`
> 2. Add tags `Co-authored-by` and `Signed-off-by` tags with your name and ideally make it full one (extend "R" to a full name)

@rmilecki, I have updated the commit as you requested (except for my full name). Please review and merge if everything looks ok. Thanks!

hauke commented:
This depends on #14738 or #14635 adding support for the flash chip.

themaverickdm commented:
> This depends on #14738 or #14635 adding support for the flash chip.

#14738 is for kernel 6.1.y, and #14635 is for 5.15.y. In any case, mediatek/filogic (WL586X3) has moved to 6.1.y as far as I can tell, so this change is only dependent on #14738.

themaverickdm commented:
@rmilecki, @hauke, the dependencies for this PR have been met, so would you please help with merging it?
Thanks.
